panther_hardware_interfaces/src/canopen_controller.cpp

```diff
@@ -0,0 +1,198 @@
+// Copyright 2023 Husarion sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <panther_hardware_interfaces/canopen_controller.hpp>
+
+#include <condition_variable>
+#include <filesystem>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <thread>
+
+#include <ament_index_cpp/get_package_share_directory.hpp>
+
+#include <realtime_tools/thread_priority.hpp>
+
+namespace panther_hardware_interfaces
+{
+
+CANopenController::CANopenController(const CANopenSettings & canopen_settings)
+: canopen_settings_(canopen_settings)
+{
+}
+
+void CANopenController::Initialize()
+{
+  if (initialized_) {
+    return;
+  }
+
+  canopen_communication_started_.store(false);
+
+  canopen_communication_thread_ = std::thread([this]() {
+    ConfigureRT();
+
+    try {
+      InitializeCANCommunication();
+    } catch (const std::system_error & e) {
+      std::cerr << "Exception caught during CAN initialization: " << e.what() << std::endl;
+
+      NotifyCANCommunicationStarted(false);
+      return;
+    }
+
+    NotifyCANCommunicationStarted(true);
+
+    try {
+      loop_->run();
+    } catch (const std::system_error & e) {
+      // If the error happens and loop stops SDO and PDO operations will timeout and in result
+      // system will switch to error state
+      std::cerr << "Exception caught in loop run: " << e.what() << std::endl;
+    }
+  });
+
+  if (!canopen_communication_started_.load()) {
+    std::unique_lock<std::mutex> lck(canopen_communication_started_mtx_);
+    canopen_communication_started_cond_.wait(lck);
+  }
+
+  if (!canopen_communication_started_.load()) {
+    throw std::runtime_error("CAN communication not initialized");
+  }
+
+  BootDrivers();
+
+  initialized_ = true;
+}
+
+void CANopenController::Deinitialize()
+{
+  // Deinitialization should be done regardless of the initialized_ state - in case some operation
+  // during initialization fails, it is still necessary to do the cleanup
+
+  if (master_) {
+    master_->AsyncDeconfig().submit(*exec_, [this]() { ctx_->shutdown(); });
+  }
+  if (canopen_communication_thread_.joinable()) {
+    canopen_communication_thread_.join();
+  }
```

Comment:
```suggestion
  if (canopen_communication_thread_ && canopen_communication_thread_.joinable()) {
    canopen_communication_thread_.join();
  }
```
What is the better approach? Should we unify this?

Reply:
I'm pretty sure that it isn't better approach, as it doesn't work :p thread has no conversion to bool, and I think that calling `joinable` is the correct way.

Reply:
Oh, okay. I thought it was a smart pointer ;-;